ouroboros-network/src/Ouroboros/Network/Protocol/TxSubmission/Type.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE EmptyCase          #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE GADTs              #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies       #-}
+
+module Ouroboros.Network.Protocol.TxSubmission.Type where
+
+import           Numeric.Natural (Natural)
+
+import           Network.TypedProtocol.Core
+
+-- |
+-- Trasaction-submission protocol states; states transitions are given by
+-- @'Message' ('TxSubmission' hash tx)@ below.
+--
+data TxSubmission hash tx where
+  -- |
+  -- Client has agency; it can either terminate, ask for transaction hashes or
+  -- ask for a transaction.
+  --
+  StIdle       :: TxSubmission hash tx
+
+  -- |
+  -- Server has agency; it must reply with a list of transaction hashes.
+  --
+  StSendHashes :: TxSubmission hash tx
+
+  -- |
+  -- Server has agency;  it must replay with a trasaction.
+  StSendTx     :: TxSubmission hash tx
+
+  -- |
+  -- Nobody has agency; termination state.
+  StDone       :: TxSubmission hash tx
+
+
+instance Protocol (TxSubmission hash tx) where
+
+  data Message (TxSubmission hash tx) from to where
+    -- |
+    -- Get a list of transaction's hashes from the server.
+    --
+    MsgGetHashes
+      :: Natural
+      -> Message (TxSubmission hash tx) StIdle StSendHashes
+
+    -- |
+    -- Reply with a list of available transaction's hashes.
+    --
+    MsgSendHashes
+      :: [hash]
+      -> Message (TxSubmission hash tx) StSendHashes StIdle
+
+    -- |
+    -- Request a transaction corresponding to the given hash.
+    --
+    MsgGetTx
+      :: hash
+      -> Message (TxSubmission hash tx) StIdle StSendTx
+
+    -- |
+    -- Send a transaction.
+    --
+    MsgTx
+      :: Maybe tx
```

Comment:
Maybe because?

Reply:
Right, if a sender announced a hash it should have it.